### PR TITLE
Verify will fail and report file/line number when called on a request not executed

### PR DESCRIPTION
**Pull request by piofusco**
`verify` now includes file and line number by default so that failures will be reported in the tests that calls them - not where we implement the function.

Additionally, I am going to leave in the return value and make it discardable. Could be valuable for users to track how many times a request was made.

Added some more examples to the README as well.

### CODE REVIEW

**Comment by jacobsimeon** [comment]
Looks good! This will definitely clean up verification for most use cases. 

Instead of returning the number of calls, let's have `verify` take a `times: Int? = nil` parameter. When it's nil, we'll just assert that the request has been received at least once. Otherwise we can assert that the request was sent the given number of times.

**Comment by piofusco** [comment]
Great idea. Updated `Verifier`

```swift
public protocol Verifier {
    func verify(_ method: Method, _ uri: String, times: UInt, file: StaticString, line: UInt)
}
```

Which now takes an `UInt`, times, which defaults to 1.

```swift
extension Application: Verifier {
    public func verify(_ method: Method, _ uri: String, times: UInt = 1, file: StaticString = #filePath, line: UInt = #line) {
        if requestsMap[RequestMade(method: method, uri: uri), default: 0] < times {
            XCTFail("Unable to verify request \(method) \(uri)\t", file: file, line: line)
        }
    }
}
```

So that we can write tests like this

```swift
func test_verify__requestsHaveBeenMade__verifierPassesTest() throws {
    subject.routes.get("/greeting") { _, _ in
        Response(status: Status.ok, content: "hola")
    }
    subject.routes.post("/tasks") { request, _ in
        Response(status: .ok)
    }

    _ = try client.get(url: "http://localhost:8080/greeting").wait()
    _ = try client.post(url: "http://localhost:8080/tasks", body: .string("build a web framework")).wait()
    _ = try client.get(url: "http://localhost:8080/greeting").wait()

    subject.verify(Method.GET, "/greeting", times: 2)
    subject.verify(Method.GET, "/greeting")
    subject.verify(Method.POST, "/tasks", times: 1)
    subject.verify(Method.POST, "/tasks")
}
```

Or conversely,

```swift
func test_verify__noRequestsHaveBeenMade__verifierFails() throws {
    subject.routes.get("/greeting") { _, _ in
        Response(status: Status.ok, content: "hola")
    }
    subject.routes.post("/tasks") { _, _ in
        Response(status: .ok)
    }

    XCTExpectFailure("The following requests were not made; verification will fail") {
        subject.verify(Method.GET, "/greeting")
        subject.verify(Method.POST, "/task")
        subject.verify(Method.GET, "/greeting", times: 100)
    }
}
```

File: Sources/Sails/Application.swift [hunk 1]
```diff
@@ -110,24 +115,29 @@ class FunctionHandler: ChannelInboundHandler {
...
+            context.write(self.wrapOutboundOut(.head(head)), promise: nil)
+            context.write(self.wrapOutboundOut(.body(.byteBuffer(buffer))), promise: nil)
 
-            context.writeAndFlush(self.wrapOutboundOut(.body(.byteBuffer(buffer)))).whenComplete { (_) in
-                _ = context.close()
+            let keepAlive = self.requestHead?.isKeepAlive == true
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil))).whenComplete { _ in
+                if !keepAlive {
+                    context.close(promise: nil)
+                }
```
Comment (jacobsimeon) [inline Sources/Sails/Application.swift]: Preferably all of these changes would be in a separate commit (looks like you pulled this stuff in from my other branch)

File: Sources/Sails/Application.swift [hunk 2]
```diff
@@ -63,8 +65,10 @@ public class Application {
 }
 
 extension Application: Verifier {
-    public func verify(_ method: Method, _ uri: String) -> Int {
-        requestsMap[RequestMade(method: method, uri: uri), default: 0]
+    public func verify(_ method: Method, _ uri: String, times: UInt = 1, file: StaticString = #filePath, line: UInt = #line) {
+        if requestsMap[RequestMade(method: method, uri: uri), default: 0] < times {
+            XCTFail("Unable to verify request \(method) \(uri)\t", file: file, line: line)
+        }
```
Comment (jacobsimeon) [inline Sources/Sails/Application.swift]: Looks good but, I was actually thinking that if we pass `times` then it would fail if the the request hasn't been sent that exact number of times. 

```swift
// This should fail if the request has been sent 1 or 3 times
server.verify(.get, "http://example.com", times: 2) 

// If we don't pass a value for `times` then it should pass as long as the request has been sent at least once
server.verify(.get, "http://example.com")
```

So I was thinking of an implementation like this:
```swift
public func verify(_ method: Method, _ uri: String, times: UInt? = nil, file: StaticString = #filePath, line: UInt = #line) {
    let callCount = requestsMap[RequestMade(method: method, uri: uri), default: 0]
    if let times = times, times != callCount {
         XCTFail("Expected to receive request \(times) times but it was sent \(callCount) times")
         return
    }
    
    if callCount < 1 {
         XCTFail("Unable to verify request \(method) \(uri)\t", file: file, line: line)
    }
}
```

I'm writing this while in a zoom meeting so the above code examples are probably not perfect

File: Sources/Sails/Application.swift (hunk 1 above)
Comment (piofusco) [reply Sources/Sails/Application.swift]: Yea, unfortunately the tests wouldn't pass without them so I wasn't sure what else to do.

File: README.md [hunk 3]
```diff
@@ -66,9 +67,18 @@ class PullToRefreshTests: XCTestCase {
...
         // Ensure that only one request was made to POST /graphql
-        XCTAssertEqual(server.verify(.POST, "/graphql"), 1)
+//        XCTAssertEqual(server.verify(.POST, "/graphql"), 1)
+    
+        // Inversely, we can also ensure requests were not made
+        XCTExpectFailure("Ensure following requests were not made") {
+            subject.verify(Method.GET, "/greeting")
+            subject.verify(Method.POST, "/task")
+        }
     }
```
Comment (jacobsimeon) [inline README.md:82]: Dig all this documentation and examples! So nice

File: Sources/Sails/Application.swift (hunk 1 above)
Comment (piofusco) [reply Sources/Sails/Application.swift]: Should I rip these out and maybe put them up in their own PR?

File: Sources/Sails/Application.swift [hunk 4]
```diff
@@ -63,8 +65,16 @@ public class Application {
 }
 
 extension Application: Verifier {
-    public func verify(_ method: Method, _ uri: String) -> Int {
-        requestsMap[RequestMade(method: method, uri: uri), default: 0]
+    public func verify(_ method: Method, _ uri: String, times: UInt? = 1, file: StaticString = #filePath, line: UInt = #line) {
```
Comment (jacobsimeon) [inline Sources/Sails/Application.swift]: default value for `times` needs to be `nil` 😬

otherwise we'll never get past that first conditional

File: Sources/Sails/Application.swift (hunk 1 above)
Comment (jacobsimeon) [reply Sources/Sails/Application.swift]: I merged my PR with these changes. You can pull this stuff out and then rebase on `master` and you should be good

**Review by jacobsimeon — APPROVED** [review]
Awesome! Thanks
